Packmol: overhaul and add tests

rkingsbury opened the pull request:
## Summary

This PR overhauls `PackmolWrapper` and adds a set of thorough unit tests. These changes were motivated by my recent experience using the pymatgen `PackmolRunner` class (which has many deficiencies) and this one to try to build structures. With this PR, I have brought the `mdgo` version up to feature parity with the pymatgen one and made it more robust to cases where packmol times out or fails to satisfy all packing constraints.

Major changes include:

* Changing the call signature so that the dict (formerly `structures` now `molecules`) also contains the numbers of each molecule. Previously the numbers were passed as a separate list. To me, if we are already providing a `[dict]` that specifies the names and structures of every molecule, we might as well include the numbers there too to minimize chances of a mistake.
* Changed the "structure" key of the input dict to "coords"
* Add several features that the pymatgen `PackmolRunner` had but this class did not. Namely, support for `Molecule` objects, ability to customize control parameters, and ability to automatically determine box size
* Added a timeout option to the `run_packmol()` command
* Added logic to catch cases where packmol fails but still returns zero exit code (see https://github.com/m3g/packmol/issues/28)
* Allow files to be given as `Path` in addition to `str`
* Support filenames or paths that contain spaces
* Add unit tests for everything

## Additional dependencies introduced (if any)

* None

## TODO - thinking longer term

I believe this PR is good to go, pending review.

_However_, longer term I also feel that it would make sense to move this functionality back into pymatgen and create a proper `InputSet` class for packmol that follows the forthcoming [abstract IO interface](https://gist.github.com/rkingsbury/8e9e076a9d0d95919d66acc93bd438df) I've been developing. For example, instead of `make_packmol_input` we would  use `write_input` to be consistent with that interface.

It doesn't really make sense to adopt that new interface within`mdgo` until it's merged into pymatgen. So one approach would be to merge this PR for use in `mdgo` now, with the understanding that it may quickly be superceded by a parallel implementation in `pymatgen` that has a slightly different interface. When that happens, I will deprecate the pymatgen `PackmolRunner` and we could also add a deprecation warning here. It's a little messy, but since not many people use `mdgo` yet I think it's OK.

An alternate approach would be to go ahead and change the interface here so that it looks and acts just like what will eventually be in `pymatgen`, without actually inheriting any of those base classes yet. That way people can start getting used to the new interface and not have to change their code when we update it to use the abcs in pymatgen later.

Happy to discuss in more detail @htz1992213 .

Before a pull request can be merged, the following items must be checked:

- [x] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/). The easiest way to handle this
      is to run the following in the **correct sequence** on your local machine. Start with running
      [black](https://black.readthedocs.io/en/stable/index.html) on your new code. This will automatically reformat
      your code to PEP8 conventions and removes most issues. Then run 
      [pycodestyle](https://pycodestyle.readthedocs.io/en/latest/), followed by 
      [flake8](http://flake8.pycqa.org/en/latest/).
- [x] Docstrings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
      Run [pydocstyle](http://www.pydocstyle.org/en/2.1.1/index.html) on your code.
- [x] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org/) to type check your code.
- [x] Tests have been added for any new functionality or bug fixes.
- [x] All linting and tests pass.

rkingsbury commented:
@htz1992213 after thinking more about this, I think there are 2 possible ways we can proceed. Either way, I'd like to develop a separate packmol IO class as part of my htmd branch in pymatgen. There's no reason this `PackmolWrapper` can't exist alongside the IO class in pymatgen; although I think eventually we should encourage people towards the full IO class once it's available.

Given that, we can either:

1. Review and merge this as is, or 
2. Modify this PR so that the call signature of `PackmolWrapper` does not change at all (i.e., go back to `structures` and `numbers` args), to facilitate backward compatibility. This would avoid people having to update their code between now and when the IO class is ready.

Please let me know which way you'd like to go. Thanks!

rkingsbury commented:
FYI @htz1992213 I have now implemented a full I/O class for packmol in pymatgen. See [this file](https://github.com/materialsproject/pymatgen/pull/2211/files?authenticity_token=rzTdjcRVFJvYZXT0oJKq7%2BaOmuOJBJatCDWaKlC%2Baehd4dX%2Ffkv6azoUELgYJcSYjuvnoq3scXs%2FLeiWnqEEJQ%3D%3D&file-filters%5B%5D=.py#diff-e4aea55798cc59cfc4ee643be53c6115ed69977fb706424b5057981204da5ca3)

It's not merged yet so still somewhat subject to change. A lot of the contents are taken from my work here, with some slight modifications like not using the mdgo method of volume estimation, for example.

htz1992213 commented:
@rkingsbury Thanks for the overhaul, the changes look great! One minor thing is about the tests. I added a packmol executable so that the CI can run the tests. The `MoleculeMatcher` in the `test_random_seed` test utilizes openbabel features which require additional dependencies and cause the CI to fail. Would it be possible to use an alternative approach to check packed structures are the same? For example by directly checking the XYZ file?

htz1992213 commented on tests/test_mdgopackmol.py:
```diff
@@ -1,11 +1,233 @@
+from subprocess import TimeoutExpired
 
+import pytest
+from pymatgen.analysis.molecule_matcher import MoleculeMatcher
```
```suggestion
import numpy as np
```

htz1992213 commented on tests/test_mdgopackmol.py:
```diff
@@ -1,11 +1,233 @@
+        while seed = 1 is deterministic
+        """
+        mm = MoleculeMatcher()
+
```
```suggestion
```

htz1992213 commented on tests/test_mdgopackmol.py:
```diff
@@ -1,11 +1,233 @@
+            out1 = Molecule.from_file(os.path.join(scratch_dir, "output.xyz"))
+            pw.run_packmol()
+            out2 = Molecule.from_file(os.path.join(scratch_dir, "output.xyz"))
+            assert mm.fit(out1, out2)
```
```suggestion
            assert np.array_equal(out1.cart_coords, out2.cart_coords)
```

htz1992213 commented on tests/test_mdgopackmol.py:
```diff
@@ -1,11 +1,233 @@
+            out1 = Molecule.from_file(os.path.join(scratch_dir, "output.xyz"))
+            pw.run_packmol()
+            out2 = Molecule.from_file(os.path.join(scratch_dir, "output.xyz"))
+            assert not mm.fit(out1, out2)
```
```suggestion
            assert not np.array_equal(out1.cart_coords, out2.cart_coords)
```

htz1992213 reviewed (commented):
@rkingsbury I have suggested changes to the test that use alternative checks. Could you please take a look? Thanks!

rkingsbury commented:
> 
> 
> @rkingsbury Thanks for the overhaul, the changes look great! One minor thing is about the tests. I added a packmol executable so that the CI can run the tests. The `MoleculeMatcher` in the `test_random_seed` test utilizes openbabel features which require additional dependencies and cause the CI to fail. Would it be possible to use an alternative approach to check packed structures are the same? For example by directly checking the XYZ file?

Thanks for the good suggestions @htz1992213 ; I've adopted all of them. In addition, per discussion over in https://github.com/rkingsbury/pymatgen/pull/3, I removed the `"` from pathnames in the input file as I received a report that they were causing problems with older version of packmol. However, it seems like removing them caused the tests here to start failing. Which version of packmol are you using the CI?

orioncohen commented:
Just wanted to chime in. I've had issues with the quotation marks in both `v20.010` and `v20.3.2`.  I'm not sure what we settled on here, but I am curious if anyone solved the quotation mark issue. I am also quite puzzled by the "quotation mark mystery".

rkingsbury commented:
> Just wanted to chime in. I've now had issues with the quotation marks in both `v20.010` and `v20.3.2`. I'm not sure what we settled on here, but I am curious if anyone solved the quotation mark issue. I am also quite puzzled by the "quotation mark mystery".

Hey @orioncohen , I _think_ Tingzheng and I figured it out. Here are the highlights as best I've been able to figure out:

1. `v20.010` from conda is actually `v20.3.1`. The version number is bad (I should open an issue about it at packmol but haven't yet)
2. Quotation marks break older versions (prior to 20.3)
3. Quotation marks break newer version when there are no spaces in the pathnames (I think?)
4. Quotation marks are required by the newer version when there are spaces in the pathnames (I think?)

That's the logic that's implemented in Tingzheng's recent fix, which allowed all tests to pass here. Is this consistent with your testing?  I'm going to add the fixes to my pymatgen version as well.

rkingsbury commented:
Made a note about the version number issue here https://github.com/m3g/packmol/issues/6
